### DISCUSSION TITLE
Connection via webrepl

### DISCUSSION THREAD

**Discussion by Josverl** [pycom/pymakr-vsc#225] [Ideas]
As the underlying mpy~control library has support for a webrepl connection, it might be useful to expose that in pymakr

One possible issue is that there is unclarity on support for binary file upload/downloads

**Comment by jakobrosenberg** [comment]
I'd love this, but is it possible? AFAIK Pycom devices support Telnet and micropython-ctl supports webrepl.

**Reply by Josverl** [reply]
I think is a question if PyMakr wants to be a tool for Pycom devices, or also support generic MicroPython devices.

Pymakr v1. telnet connections assumed that there was a specific username/password prompt that is not always there in OSS MicroPython telnet implementations  See #74 

It would be really cool if V2 supports more options:
- **serial** 
- **telnet** ( to be added to micropython-ctl(-cont) ; Pycom and other families/ports that include a stable telnet server
- **webrepl** ( already in micropython-ctl) , possibly a simple plug-in into pymakr, and would be helpful for the community 
and possibly in the future:
- webserial 
- bluetooth 


also possibly webrepl could be addedd to pycom by uploading / importing the modules, but there may be compat issues to solve 
https://github.com/micropython/micropython/tree/master/extmod/webrepl

**Reply by jakobrosenberg** [reply]
I'd love for PyMakr to be a generic tool. I also think it would help the project grow. That said, it would probably have to be an OS effort.

Do you know of any pros/cons of telnet vs webrepl?

**Reply by Josverl** [reply]
## Webrepl 
**Pro :** 
- supported by all micropython ports that have network support 
- lightweight implementation in python that should be supported on other families such as PyCom 
- includes file transfer options , ( but no listdir option) 
**Con:** 
- limited support for large/binary files 
- limited recent updates or new functionality 

## Telnet 
** Pro :** 
 - standard part of pycom, (and loboris) family of firmwares 
 - standard protocol, (but lacks formal file transfer) 
**Con:**
 - little support for telnet on other families ,
 - Often requires patches to telnet client / server to work-around username/password differences 
   https://awesome-micropython.com/#telnet

**Comment by sebi5361** [comment]
FYI: there is also this webrepl console application called [zepl](https://gitlab.com/zepl1/zepl) you might want to get inspired from?!

**Comment by daniel-da** [comment]
I've implemeted the webrepl connection on PR#283

And for more stability there is another PR in micropython-ctl project waiting for review

Regards,
Daniel
